File: dotnet-backend/CanchasSinteticas.Domain/ValueObjects/TimeSlot.cs
using CanchasSinteticas.Domain.Exceptions;

namespace CanchasSinteticas.Domain.ValueObjects;

/// <summary>
/// Franja de tiempo de una reserva en un día concreto. Value object inmutable.
/// Las reglas de horario de operación y duración de bloque dependen de la sede/cancha
/// y se validan en la capa de aplicación; aquí solo se garantiza un rango coherente.
/// </summary>
public class TimeSlot
{
    /// <summary>Fecha del turno.</summary>
    public DateOnly Date { get; }

    /// <summary>Hora de inicio.</summary>
    public TimeOnly StartTime { get; }

    /// <summary>Hora de fin.</summary>
    public TimeOnly EndTime { get; }

    /// <summary>Crea una franja validando que el rango sea coherente.</summary>
    public TimeSlot(DateOnly date, TimeOnly startTime, TimeOnly endTime)
    {
        if (endTime <= startTime)
            throw new InvalidBlockError();

        Date = date;
        StartTime = startTime;
        EndTime = endTime;
    }

    /// <summary>Duración de la franja.</summary>
    public TimeSpan Duration => EndTime - StartTime;

    /// <summary>Fecha y hora de inicio combinadas.</summary>
    public DateTime StartDateTime => Date.ToDateTime(StartTime);

    /// <summary>Fecha y hora de fin combinadas.</summary>
    public DateTime EndDateTime => Date.ToDateTime(EndTime);

    /// <summary>Indica si la franja respeta la anticipación mínima requerida.</summary>
    public bool IsBookable(DateTime now, int minAdvanceMinutes) =>
        StartDateTime - now >= TimeSpan.FromMinutes(minAdvanceMinutes);

    /// <summary>Indica si la franja está dentro del horario de operación indicado.</summary>
    public bool WithinOperatingHours(TimeOnly opening, TimeOnly closing) =>
        StartTime >= opening && EndTime <= closing;

    /// <summary>Indica si esta franja se solapa con otra.</summary>
    public bool OverlapsWith(TimeSlot other)
    {
        if (Date != other.Date)
            return false;
        return StartTime < other.EndTime && other.StartTime < EndTime;
    }
}

File: backend/domain/value_objects/time_slot.py
from dataclasses import dataclass
from datetime import date, time, datetime, timedelta

from domain.exceptions import DurationError, InvalidBlockError, OperatingHoursError

_OPERATING_START = time(6, 0)
_OPERATING_END = time(23, 0)
_VALID_MINUTES = {0, 30}


@dataclass
class TimeSlot:
    date: date
    start_time: time
    end_time: time

    def __post_init__(self) -> None:
        if self.start_time.minute not in _VALID_MINUTES or self.end_time.minute not in _VALID_MINUTES:
            raise InvalidBlockError()

        start_dt = datetime.combine(self.date, self.start_time)
        end_dt = datetime.combine(self.date, self.end_time)
        if end_dt - start_dt < timedelta(hours=1):
            raise DurationError()

        if self.start_time < _OPERATING_START or self.end_time > _OPERATING_END:
            raise OperatingHoursError()

    @property
    def start_datetime(self) -> datetime:
        return datetime.combine(self.date, self.start_time)

    @property
    def end_datetime(self) -> datetime:
        return datetime.combine(self.date, self.end_time)

    def is_bookable(self, now: datetime) -> bool:
        return self.start_datetime - now >= timedelta(hours=1)

    def overlaps_with(self, other: "TimeSlot") -> bool:
        if self.date != other.date:
            return False
        return self.start_time < other.end_time and other.start_time < self.end_time
